File: src/codes/single-state-handling/php/rfc_state.php
<?php

interface DocumentState {
  public function publish(DocumentContext $context): void;
}

class DraftState implements DocumentState {
  public function publish(DocumentContext $context): void {
    echo "レビュー依頼を送信\n";
    $context->setState(new ReviewState());
  }
}

class ReviewState implements DocumentState {
  public function publish(DocumentContext $context): void {
    echo "公開しました\n";
    $context->setState(new PublishedState());
  }
}

class PublishedState implements DocumentState {
  public function publish(DocumentContext $context): void {
    echo "すでに公開済みです\n";
  }
}

class DocumentContext {
  private DocumentState $state;

  public function __construct() {
    $this->state = new DraftState();
  }

  public function setState(DocumentState $state): void {
    $this->state = $state;
  }

  public function publish(): void {
    $this->state->publish($this);
  }
}

// 利用例
$doc = new DocumentContext();
$doc->publish(); // → レビュー依頼
$doc->publish(); // → 公開
$doc->publish(); // → すでに公開済み

File: src/codes/single-state-handling/py/rfc_state.py
from abc import ABC, abstractmethod

class DocumentState(ABC):
    @abstractmethod
    def publish(self, context: "DocumentContext"):
        pass

class DraftState(DocumentState):
    def publish(self, context: "DocumentContext"):
        print("レビュー依頼を送信")
        context.set_state(ReviewState())

class ReviewState(DocumentState):
    def publish(self, context: "DocumentContext"):
        print("公開しました")
        context.set_state(PublishedState())

class PublishedState(DocumentState):
    def publish(self, context: "DocumentContext"):
        print("すでに公開済みです")

class DocumentContext:
    def __init__(self):
        self._state: DocumentState = DraftState()

    def set_state(self, state: DocumentState):
        self._state = state

    def publish(self):
        self._state.publish(self)

# 利用例
doc = DocumentContext()
doc.publish()  # → レビュー依頼
doc.publish()  # → 公開
doc.publish()  # → すでに公開済み
